### PR TITLE
Create SBOMs for images pulled during package creation

### LINKED ISSUE #22: Produce SBOM during zarf package creation
Evaluate tools / options to produce a standard consumable SBOM and a user-friendly display / ability to navigate it when deploying a zarf package.  Doing so will meet software supply chain requirements for government and industry as well as provide a higher level of transparency/confidence to what is transported with Zarf.

Related:
https://www.ntia.doc.gov/files/ntia/publications/sbom_minimum_elements_report.pdf
https://cyclonedx.org/
https://spdx.dev/
https://github.com/anchore/syft

---

Migrated from https://repo1.dso.mil/platform-one/big-bang/apps/product-tools/zarf/-/issues/35

### PR DESCRIPTION
**Pull request by mikhailswift**
## Description

Creates SBOMs for all images pulled during `zarf package create`. Stores them in a sboms directory within the package.

## Related Issue

Fixes #22

## Type of change

- [x] New feature (non-breaking change which adds functionality)
## Checklist before merging

- [ ] Tests have been added/updated as necessary (add the `needs-tests` label)
- [ ] Documentation has been updated as necessary (add the `needs-docs` label)
- [x] An ADR has been written as necessary (add the `needs-adr` label) [ [1](https://github.com/joelparkerhenderson/architecture-decision-record) [2](https://cognitect.com/blog/2011/11/15/documenting-architecture-decisions) [3](https://adr.github.io/) ]
- [x] (Optional) Changes have been linted locally with [golangci-lint](https://github.com/golangci/golangci-lint). (NOTE: We haven't turned on lint checks in the pipeline yet so linting may be hard if it shows a lot of lint errors in places that weren't touched by changes. Thus, linting is optional right now.)

### CODE REVIEW

**Comment by mikhailswift** [comment]
Relevant pr: https://github.com/testifysec/witness/pull/167

File: go.mod [hunk 1]
```diff
@@ -5,7 +5,6 @@ go 1.16
 require (
```
[the remark this answers is not in the document]
Comment (jeff-mccoy) [reply go.mod:5]: We should open a new issue for go update but I'm really just eagerly waiting for .18 and my beautiful generics 😍

[the remark this answers is not in the document]
Comment (mikhailswift) [reply go.mod:5]: Newly imported code has modules that require go >= 1.17

[the remark this answers is not in the document]
Comment (mikhailswift) [reply go.mod:5]: Namely:

```
# sigs.k8s.io/json/internal/golang/encoding/json
/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20211208200746-9f7c6b3444d2/internal/golang/encoding/json/encode.go:1249:12: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20211208200746-9f7c6b3444d2/internal/golang/encoding/json/encode.go:1255:18: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
note: module requires Go 1.17
```

[the remark this answers is not in the document]
Comment (mikhailswift) [reply go.mod:5]: And upgrading the zarf module to 1.17 causes unhappiness with opentelemetry 

```
go: finding module for package go.opentelemetry.io/otel/metric/registry
go: finding module for package go.opentelemetry.io/otel/semconv
github.com/defenseunicorns/zarf/test/e2e imports
	github.com/rancher/k3d/v5/cmd/cluster imports
	github.com/rancher/k3d/v5/pkg/client imports
	github.com/rancher/k3d/v5/pkg/runtimes/docker imports
	github.com/docker/cli/cli/command imports
	github.com/docker/cli/cli/trust imports
	github.com/theupdateframework/notary/trustpinning tested by
	github.com/theupdateframework/notary/trustpinning.test imports
	github.com/cloudflare/cfssl/helpers imports
	github.com/google/certificate-transparency-go imports
	go.etcd.io/etcd/v3 imports
	go.etcd.io/etcd/tests/v3/integration imports
	go.etcd.io/etcd/server/v3/embed imports
	go.opentelemetry.io/otel/semconv: module go.opentelemetry.io/otel@latest found (v1.4.1), but does not contain package go.opentelemetry.io/otel/semconv
github.com/defenseunicorns/zarf/test/e2e imports
	github.com/rancher/k3d/v5/cmd/cluster imports
	github.com/rancher/k3d/v5/pkg/client imports
	github.com/rancher/k3d/v5/pkg/runtimes/docker imports
	github.com/docker/cli/cli/command imports
	github.com/docker/cli/cli/trust imports
	github.com/theupdateframework/notary/trustpinning tested by
	github.com/theupdateframework/notary/trustpinning.test imports
	github.com/cloudflare/cfssl/helpers imports
	github.com/google/certificate-transparency-go imports
	go.etcd.io/etcd/v3 imports
	go.etcd.io/etcd/tests/v3/integration imports
	go.etcd.io/etcd/server/v3/embed imports
	go.opentelemetry.io/otel/exporters/otlp imports
	go.opentelemetry.io/otel/sdk/metric/controller/basic imports
	go.opentelemetry.io/otel/metric/registry: module go.opentelemetry.io/otel/metric@latest found (v0.27.0), but does not contain package go.opentelemetry.io/otel/metric/registry
```

Which is addressed by the opentelemetry team here:

https://github.com/open-telemetry/opentelemetry-go/issues/2422#issuecomment-981716733

But... it seems like that's going to require changes somewhere in that dependency change, probably etcd

[the remark this answers is not in the document]
Comment (jeff-mccoy) [reply go.mod:5]: Okay cool, let's just bump to 1.17 everywhere then properly, no issues going to 1.17 for me.

[the remark this answers is not in the document]
Comment (mikhailswift) [reply go.mod:5]: Alright, I think I've managed to escape go mod hell and get the weird dependency issues sorted now. Should be bumped to 1.17 now

File: cli/internal/packager/create.go [hunk 2]
```diff
@@ -52,7 +53,9 @@ func Create() {
 
 	if len(seedImages) > 0 {
 		// Load seed images into their own happy little tarball for ease of import on init
-		images.PullAll(seedImages, tempPath.seedImages)
+		pulledImages := images.PullAll(seedImages, tempPath.seedImages)
+		_ = utils.CreateDirectory(tempPath.sboms, 0700)
+		sbom.CatalogImages(pulledImages, tempPath.sboms)
```
Comment (YrrepNoj) [inline cli/internal/packager/create.go:58]: @jeff-mccoy This is writing the sboms to the temp directory which I believe gets blown away when the package creation is finished (unless there is a fatal error). Is there a persistent but intuitive place you think these should be written to instead?

Comment (mikhailswift) [reply cli/internal/packager/create.go:58]: From my testing the sboms directory is getting packaged into the .zst

Comment (YrrepNoj) [reply cli/internal/packager/create.go:58]: Oops you're totally correct, I forgot we archive the temp directory into the package itself.

**Review by YrrepNoj — DISMISSED** [review]

**Comment by jeff-mccoy** [comment]
This is going to need some more work before we can merge it.  We will need to have an ADR on the libs we pulled in (it adds multiple deps and increases the binary size by > 10%) as well as a test to verify it's being created.  Also not sure what's going on with the `go.mod`, but it's enormous compared to master.  Finally, we'll need some doc entry to explain what users should see.  ADR won't be needed after the first SBOM PR and we can help with that if needed.

File: cli/internal/sbom/catalog.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/testifysec/witness/pkg/attestation"
+	"github.com/testifysec/witness/pkg/attestation/syft"
+)
+
+func CatalogImages(tagToImage map[name.Tag]v1.Image, sbomDir string) {
+	imageCount := len(tagToImage)
+	spinner := message.NewProgressSpinner("Creating SBOMs for %d images.", imageCount)
+	actx, err := attestation.NewContext([]attestation.Attestor{})
```
Comment (jeff-mccoy) [inline cli/internal/sbom/catalog.go:20]: as a pattern, we always run `defer spinner.Stop()` immediately after invoking `message.NewProgressSpinner()` to ensure there are no spinner leaks.

File: go.mod [hunk 4]
```diff
@@ -1,37 +1,266 @@
...
 	k8s.io/klog/v2 v2.40.1
 	sigs.k8s.io/kind v0.11.1
-	sigs.k8s.io/kustomize/api v0.8.11
-	sigs.k8s.io/kustomize/kyaml v0.11.0
+	sigs.k8s.io/kustomize/api v0.10.1
+	sigs.k8s.io/kustomize/kyaml v0.13.0
 	sigs.k8s.io/yaml v1.3.0
 )
+
+require (
```
Comment (jeff-mccoy) [inline go.mod:36]: Need to understand what is happening with all these explicit indirect deps.

File: go.mod [hunk 5]
```diff
@@ -1,37 +1,266 @@
...
 	github.com/spf13/cobra v1.3.0
 	github.com/stretchr/testify v1.7.0
-	golang.org/x/crypto v0.0.0-20211215165025-cf75a172585e
+	github.com/testifysec/witness v0.1.7-0.20220307231058-f11e00e711dd
+	golang.org/x/crypto v0.0.0-20220213190939-1e6e3497d506
 	helm.sh/helm/v3 v3.7.2
-	k8s.io/api v0.22.5
-	k8s.io/apimachinery v0.22.5
-	k8s.io/client-go v0.22.5
+	k8s.io/api v0.23.4
```
Comment (jeff-mccoy) [inline go.mod:26]: whenever we jump k8s versions we need to make sure we're careful to not introducing any version issues with remote clusters, was this bump needed for SBOM or just arbitrary?

File: go.mod [hunk 6]
```diff
@@ -1,37 +1,266 @@
...
 	github.com/mattn/go-colorable v0.1.12
 	github.com/mholt/archiver/v3 v3.5.1
-	github.com/mitchellh/go-wordwrap v1.0.1 // indirect
 	github.com/otiai10/copy v1.7.0
 	github.com/pterm/pterm v0.12.33
 	github.com/rancher/k3d/v5 v5.2.1
 	github.com/spf13/cobra v1.3.0
 	github.com/stretchr/testify v1.7.0
-	golang.org/x/crypto v0.0.0-20211215165025-cf75a172585e
+	github.com/testifysec/witness v0.1.7-0.20220307231058-f11e00e711dd
```
Comment (jeff-mccoy) [inline go.mod:23]: Would be good to understand the witness versioning plan and how that will be leveraged by zarf in your mind

File: go.mod [hunk 7]
```diff
@@ -1,37 +1,266 @@
...
-	k8s.io/apimachinery v0.22.5
-	k8s.io/client-go v0.22.5
+	k8s.io/api v0.23.4
+	k8s.io/apimachinery v0.23.4
+	k8s.io/client-go v0.23.4
 	k8s.io/klog/v2 v2.40.1
 	sigs.k8s.io/kind v0.11.1
-	sigs.k8s.io/kustomize/api v0.8.11
-	sigs.k8s.io/kustomize/kyaml v0.11.0
+	sigs.k8s.io/kustomize/api v0.10.1
```
Comment (jeff-mccoy) [inline go.mod:31]: no issue with kustomize bumps if everything still is happy, but curious was this needed for sbom or not?

File: go.mod [hunk 8]
```diff
@@ -1,37 +1,266 @@
...
+	k8s.io/kube-openapi v0.0.0-20220124234850-424119656bbf // indirect
+	k8s.io/kubectl v0.22.4 // indirect
+	k8s.io/metrics v0.22.4 // indirect
+	k8s.io/utils v0.0.0-20220127004650-9b3446523e65 // indirect
+	oras.land/oras-go v0.4.0 // indirect
+	sigs.k8s.io/json v0.0.0-20211208200746-9f7c6b3444d2 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.1 // indirect
+)
+
+replace github.com/derailed/popeye => github.com/testifysec/popeye v0.9.9-0.20220307190025-d441c7496b07
```
Comment (jeff-mccoy) [inline go.mod:266]: Popeye is a useful tool and is used by k9s that we vendor in, running on a fork might become a real problem down the road, can you clarify the need for this?

**Review by jeff-mccoy — CHANGES_REQUESTED** [review]
As already stated, will need ADR/Docs/Tests and items discussed below.  I also noticed SBOM generation can take quite a while, any comments on performance or possible future improvements for that?

File: go.mod (hunk 4 above)
Comment (mikhailswift) [reply go.mod:36]: Go 1.17 changed the go.mod file to include indirect dependencies: https://go.dev/doc/go1.17 under heading "Pruned module graphs in go 1.17 modules"

Comment (jeff-mccoy) [reply go.mod:36]: TIL 😂 thx.

File: go.mod (hunk 5 above)
Comment (mikhailswift) [reply go.mod:26]: Needed for SBOM -- or at least currently.  Witness imports syft which imports k8s 0.23.0.  I fought with dependencies for a while trying to limit what changed.

File: go.mod (hunk 6 above)
Comment (mikhailswift) [reply go.mod:23]: Witness will follow SemVer.  We're a bit loose with breaking changes right now due to being in 0.x, which SemVer allow, but we are still making efforts to limit breaking changes.

File: go.mod (hunk 7 above)
Comment (mikhailswift) [reply go.mod:31]: Same as the answer about the k8s api

File: go.mod (hunk 8 above)
Comment (mikhailswift) [reply go.mod:266]: A field in a struct in the k8s API changed names, breaking compilation.  It's my goal to get the change upstreamed to popeye and the fork be temporary.

File: cli/internal/sbom/catalog.go (hunk 3 above)
Comment (mikhailswift) [reply cli/internal/sbom/catalog.go:20]: Can make that change -- I copied the pattern I saw in images.PullAll

Comment (jeff-mccoy) [reply cli/internal/sbom/catalog.go:20]: Yeah we have some areas we may have missed that pattern, thx.

**Comment by mikhailswift** [comment]
> As already stated, will need ADR/Docs/Tests and items discussed below. I also noticed SBOM generation can take quite a while, any comments on performance or possible future improvements for that?

I've been working on the adr/docs/test, will get something in hsortly.

I can do some tracing to see what part of the process is taking time and identify areas for improvement.

**Comment by jeff-mccoy** [comment]
That would be super helpful, thanks!

**Comment by mikhailswift** [comment]
Ran some quick tests today profiling sbom generation on `package-example-single-big-bang-package`

Here's a flame graph of the sbom code on a run with a cold image cache

image

The total run time of sbom.CatalogImages was 11.19 seconds.

And here's a flame graph for the same package but with a hot cache

image

The total run time of sbom.CatalogImages was 4.57 seconds.

From the flame graphs it appears that it's rebuilding the image tar from the layer cache instead of just using the cache directly as I hoped it would.  This explains why sbom generation is much faster on subsequent runs for the same image since the tar is already built.

Simplest fix will be to try to get the cataloger to use the layer cache directories directly instead of re-building the image's tar file.

**Comment by mikhailswift** [comment]
I spent the last bit chasing down performance improvements.  I'll get the ADR and documentation in asap this week while I continue working tests and performance issues.

**Comment by jeff-mccoy** [comment]
thanks @mikhailswift note there are some merge conflicts with the tests updates we've merged in recently.

**Review by YrrepNoj — DISMISSED** [review]
Code changes looks good to me. 

There was a single comment @jeff-mccoy left on adding a quick defer statement for the spinner that's left.

File: docs/adr/0004-generate-sboms-with-witness.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# 4. SBOM Generation with Witness
+
+Date: 2022-03-29
+
+## Status
+
+Proposed
```
Comment (jeff-mccoy) [inline docs/adr/0004-generate-sboms-with-witness.md:7]: Let's make this accepted now, thx

**Comment by mikhailswift** [comment]
Sounds good!

I've been looking into the usage of the sbom a bit, will probably have some questions on the UX of looking at these from Zarf in the short term.

**Comment by jeff-mccoy** [comment]
Overall this PR seems reasonable after messing with it for a couple of days, but I have concerns with adding SBOM and slowing down the package create stage without some visible utility to the user.  So I propose we finish and merge this ASAP but not cut a release until we add something to view/use the SBOM.  Here's something I've been playing around with the past couple of days that consumes our SBOM JSON data: https://www.loom.com/share/d85b02b13df442d7a0bae012307db3ee.  Thoughts @mikhailswift?

File: docs/adr/0004-generate-sboms-with-witness.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+Accepted
+
+## Context
+
+SBOM are required for software running on government hardware per EO14028.
+
+## Decision
+
+Using Witness' Syft attestor functionality allows Zarf to continue to get more refined SBOM capabilities as Witness' capabilities expand over time. Syft is capable of finding installed packages and some binaries for statically compiled dependencies over each image within a Zarf package. This allows for SBOMs for each image to be generated and packaged along with the Zarf package.  Capabilities to display SBOMs for a Zarf package are forthcoming.
```
Comment (jeff-mccoy) [inline docs/adr/0004-generate-sboms-with-witness.md:15]: "apabilities to display SBOMs for a Zarf package are forthcoming.", would love for that to have more info.

File: docs/adr/0004-generate-sboms-with-witness.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+## Consequences
+
+Syft has some dependencies that cause some issues currently.  Namely bumping to k8s-api v0.23 causes some incompatibilities with derailed/popeye which has resulted in a fork being made.  This bump may also be undesirable due to compatibility with clusters that do not use this version of the k8s api.  TestifySec is currently investigating removing the Syft dependency entirely and generating the SBOM without Syft which will alleviate these concerns and conflicts.
```
Comment (jeff-mccoy) [inline docs/adr/0004-generate-sboms-with-witness.md:19]: This needs to be updated now I believe.

**Comment by jeff-mccoy** [comment]
This still needs docs/tests, any updates on your end @mikhailswift?

**Comment by jeff-mccoy** [comment]
BB Core SBOM, it runs--it's also > 200 MBs 😱🤣

Screen Shot 2022-04-12 at 7 28 21 AM

**Comment by mikhailswift** [comment]
> BB Core SBOM, it runs--it's also > 200 MBs screamrofl
> 
> Screen Shot 2022-04-12 at 7 28 21 AM

wew.  something i was considering at the grocery store today is that we could potentially generate the sboms on demand instead of packaging them if the size becomes an issue.  main problem there is that currently syft doesn't support multi-image tar files, and of course it may not be an amazing user experience for the consumer of the sbom.

**Comment by jeff-mccoy** [comment]
TIL some ppl daydream about SBOM while getting groceries...

**Comment by mikhailswift** [comment]
> TIL some ppl daydream about SBOM while getting groceries...

Send help please

**Comment by mikhailswift** [comment]
Okay, I'm doing some last rounds of testing but I believe I've performance of generating the SBOMs improved significantly:

Both of the below timings were measured with .zarf-image-cache erased for a cold run of the single-big-bang-package exampoe

Current code:

```make package-example-single-big-bang-package  25.37s user 4.31s system 38% cpu 1:16.58 total```

Not a precise measurement but watching the timer on the SBOM step seemed to take ~30 seconds 

Changed:

```make package-example-single-big-bang-package  23.31s user 3.43s system 57% cpu 46.313 total```

Same imprecise measurement as above but the SBOM step seemed to take ~10 seconds.

Image pull on both examples took about 30 seconds.

Looking at cpuprofile comparisons again,

Current code:

image

In this case `image.Read` is taking up ~30% of the cpu time in the CatalogImages function call, and there are some sneaky http and tls calls in there that lead me to believe it could actually be pulling layers over again.

Changed code:
image

In this case image.Read is taking ~22% of the cpu time in CatalogImages, but no tls calls happen.

Will have this + docs & tests pushed today

**Comment by mikhailswift** [comment]
added some basic docs and updates to the adr.

worked a bit on presentation: https://github.com/defenseunicorns/zarf/commit/1d5d02f0e4533ed5e0df1c9a5f63940234cdecb2

That is very much a hacky work in progress, though

**Review by jeff-mccoy — APPROVED** [review]

**Comment by jeff-mccoy** [comment]
@Madeline-UX
Latest update with the SBOM Viewer thangz

Screen Shot 202
2-04-26 at 1 51 51 PM
 
Screen Shot 2022-04-26 at 1 54 00 PM

File: src/internal/sbom/viewer/template.gohtml [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
...7655C59.2877 42.7566 49.5866 38.6938 39.4048 38.345C26.7801 29.7203 23.2356 12.2024 22.5885 8.19345C60.1339 12.2621 69.0946 32.3266 69.1244 32.3764L69.443 33.0628L70.0304 32.6549C80.1598 26.1008 92.0984 22.902 104.151 23.5129C122.411 24.3983 136.459 33.3612 139.317 35.9874C146.694 42.5131 150.896 51.7545 152.34 61.0059C153.903 71.1128 152.15 81.2695 147.421 88.0638C144.123 92.7114 139.724 96.4707 134.617 99.0063C125.338 103.731 113.808 104.975 102.478 103.672C91.1479 102.369 80.0465 98.3696 71.6135 92.4805C66.6178 89.1791 62.4703 84.7489 59.5066 79.5485C60.1312 78.9162 60.7228 78.2522 61.2788 77.559C64.0566 74.2464 65.0523 71.1427 65.0523 69.6505C65.0611 69.3558 65.0138 69.0621 64.9129 68.7851C64.0666 66.7159 68.7361 62.9159 68.7461 62.906L69.3136 62.4484L68.8058 61.9311C68.8058 61.9311 64.5445 57.5441 64.0566 55.5247C63.6484 53.7242 69.1941 49.2975 69.214 49.2875L69.5924 48.9791L69.4132 48.5215C69.4132 48.5215 67.85 44.6917 68.2084 43.3388C68.5669 41.9859 73.9035 40.2948 73.9134 40.2948C70.4884 38.723 67.0336 42.6026 66.944 43.0105C66.5756 44.443 67.6011 47.4273 68.0292 48.5613C67.7125 48.8283 67.2286 49.2427 66.6752 49.7655C66.6719 49.7687 66.6685 49.7718 66.6652 49.775H66.6851Z"
```
Comment (YrrepNoj) [inline src/internal/sbom/viewer/template.gohtml:24]: out of curiosity, how was this file built/generated?

File: src/internal/packager/deploy.go [hunk 13]
```diff
@@ -62,9 +63,31 @@ func Deploy() {
...
+	sbomViewFiles, _ := filepath.Glob(tempPath.sboms + "/sbom-viewer-*")
+	// If SBOM files exist, temporary place them in the deploy directory
+	if len(sbomViewFiles) > 0 {
+		sbomDir := "zarf-sbom"
+		// Cleanup any failed prior removals
+		_ = os.RemoveAll(sbomDir)
+		// Create the directory again
+		utils.CreateDirectory(sbomDir, 0755)
+		for _, file := range sbomViewFiles {
+			// Our file copy lib explodes on these files for some reason...
```
Comment (YrrepNoj) [inline src/internal/packager/deploy.go:75]: This works so I guess that's fine but I'm curious as to why the copy wouldn't work.. Probably something that's worth  looking into when we have more time.

**Review by YrrepNoj — APPROVED** [review]

File: src/internal/packager/deploy.go (hunk 13 above)
Comment (jeff-mccoy) [reply src/internal/packager/deploy.go:75]: Yeah it's a bug in that library. We actually only introduced it for the recursive copy feature we really don't use anymore so I'd like to look at moving everything to this pattern (standard lib) and drop that dep soon.

File: src/internal/sbom/viewer/template.gohtml (hunk 12 above)
Comment (jeff-mccoy) [reply src/internal/sbom/viewer/template.gohtml:24]: what do you mean? It's just plain html.  You mean the embedded SVG?

**Pull request merged**